### PR TITLE
Entity listener utilities for ktx-ashley; typos

### LINKED ISSUE #409: EntityListener utilities for Ashley
- Add an `EntityRemovalListener` for removal-only
- Add an `EntityAdditionListener` for addition-only
- Kotlin lambda utils for similar functionality

### PR DESCRIPTION
**Pull request by RedGrapefruit09**
Resolves #409 .

- [x] `EntityAdditionListener`
- [x] `EntityRemovalListener`
- [x] Lambda utilities
- [x] KDocs
- [x] Tests
- [x] Changelog and `README.md` updates

### CODE REVIEW

File: ashley/src/main/kotlin/ktx/ashley/entity_listeners.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package ktx.ashley
```
Comment (czyzby) [inline ashley/src/main/kotlin/ktx/ashley/entity_listeners.kt:1]: Not sure if underscores are the standard in Kotlin file names. You can rename the file to just `listeners.kt`.

**Comment by RedGrapefruit09** [comment]
@czyzby , is this the proper way of implementing your suggestion with lambdas? (below)

```kotlin
inline fun EntitySystem.onEntityAdded(action: (entity: Entity) -> Unit) {
  engine.addEntityListener(/* some impl of EntityAdditionListener that takes a lambda? */ null)
}
```

File: ashley/src/main/kotlin/ktx/ashley/listeners.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package ktx.ashley
+
+import com.badlogic.ashley.core.Entity
+import com.badlogic.ashley.core.EntityListener
+
+/**
+ * An [EntityListener] only listening to the [entityAdded] event.
+ *
+ * A [Unit] implementation is provided for the [entityRemoved] event, as it is ignored.
```
Comment (czyzby) [inline ashley/src/main/kotlin/ktx/ashley/listeners.kt:9]: `[Unit] implementation` might not be clear for the users. I'd go with a `no-op` or `empty implementation`.

**Comment by RedGrapefruit09** [comment]
> If that's OK, please also update the changelog.

Done! I also updated the docs and added examples in the `README.md` of the `ktx-ashley` module.
I'll do this again once the lambda-related changes are finished. Thanks for your feedback, once again!

File: ashley/src/main/kotlin/ktx/ashley/listeners.kt (hunk 2 above)
Comment (RedGrapefruit09) [reply ashley/src/main/kotlin/ktx/ashley/listeners.kt]: Fixed in https://github.com/libktx/ktx/pull/410/commits/a2cdbcdf9ebcd52bdb467d5ad43b43e94096e964!
I chose to go with `empty implementation` as it seems less ambiguous to me.

File: ashley/src/main/kotlin/ktx/ashley/entity_listeners.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import com.badlogic.ashley.core.EntityListener
+
+/**
+ * An [EntityListener] only listening to the [entityAdded] event.
+ *
+ * A [Unit] implementation is provided for the [entityRemoved] event, as it is ignored.
+ *
+ * If you need both events from the [EntityListener], implement the regular [EntityListener] from Ashley.
+ */
+interface EntityAdditionListener : EntityListener {
```
[the remark this answers is not in the document]
Comment (RedGrapefruit09) [reply ashley/src/main/kotlin/ktx/ashley/entity_listeners.kt]: Done in https://github.com/libktx/ktx/pull/410/commits/9d676a84ed2bc75b96aa211bcde09108bd22cb6b!

**Comment by RedGrapefruit09** [comment]
I accidentally put the changelog changes in the `1.10.0-rc1` header, not in the `1.10.0-SNAPSHOT` header. I fixed it in https://github.com/libktx/ktx/pull/410/commits/951e9e777c91065d95e7004b26e30d0305c491a6.

File: ashley/src/test/kotlin/ktx/ashley/listenerTest.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package ktx.ashley
+
+import com.badlogic.ashley.core.Entity
+
+@Suppress("unused")
+class `should implement EntityAdditionListener with only one method implemented` : EntityAdditionListener {
+  // Guarantees EntityAdditionListener's entityRemoved method is always optional to implement
+
+  override fun entityAdded(entity: Entity?) = Unit
```
[the remark this answers is not in the document]
Comment (RedGrapefruit09) [reply ashley/src/test/kotlin/ktx/ashley/listenerTest.kt:9]: I'm not sure if the `entity` is always not null, for now, I'll change the definition in the interfaces to be `Entity?` in the next commit.

**Comment by RedGrapefruit09** [comment]
> Change the method signature to return an `EntityListener` (or even an `EntityAdditionListener`/`EntityRemovalListener`). Return the created instance of the listener, allowing the user to save a reference to it and eventually remove it from the engine.

Ok, I'll do that.

**Comment by RedGrapefruit09** [comment]
I made a prototype of the lambda utils (see https://github.com/libktx/ktx/pull/410/commits/cd8331ee7060525a770b85fd46a05527983415fc), but I, for some reason, I can't do this:

```kotlin
  // this compiles
  engine.onEntityAdded({
    
  })
  
  // this doesn't compile
  engine.onEntityAdded {
    
  }
```

**Comment by czyzby** [comment]
You have to put the lambda as the last parameter, after the family and priority.

File: ashley/src/test/kotlin/ktx/ashley/listenerTest.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (RedGrapefruit09) [reply ashley/src/test/kotlin/ktx/ashley/listenerTest.kt:9]: Resolved in new commits.

**Comment by RedGrapefruit09** [comment]
@czyzby I've finished implementing the lambda utils and added KDocs.
Are there any changes left to be made until I've got to update the changelog and README?

**Comment by czyzby** [comment]
As far as I can tell, there are only some issues related to the naming of variables and such. `emptyFamily` sounds more like `anyFamily` or `allFamilies`, since it matches every family as far as I can tell. I'd also rename the `implementation` lambda parameters to something like `listener` so that they are clearer if someone decides to use the named parameters.

**Comment by RedGrapefruit09** [comment]
@czyzby I've implemented your feedback in https://github.com/libktx/ktx/pull/410/commits/0625d710a45121ba6e52989e8130d7382a70f361.
I'll make some tests (`ktx-ashley` seems to use this Spek thing, so I guess I'll learn it) later.

**Comment by RedGrapefruit09** [comment]
I've also now updated the changelog and added some docs.

**Comment by czyzby** [comment]
About the tests: `ktx-ashley` was contributed as a whole with Spek tests and I didn't see an issue with that at the time, even though most modules use JUnit. Nowadays, I'd prefer if all tests used the same technology (and structure) throughout the project, even if that means going with a less modern approach.

If you have trouble writing Spek tests, I'll accept the PR with JUnit test suite as well. However, Spek would be more consistent with the current state of this module. I might consider rewriting the tests to JUnit some time, even though I personally don't have any problems with the Spek framework - it's just easier to rewrite 1 module for consistency rather than everything else.

**Comment by RedGrapefruit09** [comment]
I've made Spek tests that (I think) cover everything, so I think this PR is ready for review.

File: ashley/src/main/kotlin/ktx/ashley/listeners.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * An [EntityListener] only listening to the [entityAdded] event.
+ *
+ * An empty implementation is provided for the [entityRemoved] event, as it is ignored.
+ *
+ * If you need both events from the [EntityListener], implement the regular [EntityListener] from Ashley.
+ */
+interface EntityAdditionListener : EntityListener {
+  override fun entityAdded(entity: Entity)
```
Comment (czyzby) [inline ashley/src/main/kotlin/ktx/ashley/listeners.kt:19]: The only thing that comes to mind at this point is that we might want to include some documentation for the methods that should be overridden, i.e. something along the lines of `Invoked when an matching [Entity] is added to the engine. See [EntityListener.entityAdded].`.

**Review by czyzby — APPROVED** [review]
Looks great, thanks for your contribution. I'll go through this again and merge it later today.

File: ashley/src/main/kotlin/ktx/ashley/listeners.kt (hunk 5 above)
Comment (RedGrapefruit09) [reply ashley/src/main/kotlin/ktx/ashley/listeners.kt:24]: Resolved in https://github.com/libktx/ktx/pull/410/commits/56811de1b8d027612a7422f7a36cc53205a68cf5.

**Comment by czyzby** [comment]
Thanks for the contribution, @RedGrapefruit09!

**Pull request merged**